ratatool-diffy/src/main/scala/com/spotify/ratatool/diffy/ProtoBufDiffy.scala

```diff
@@ -59,40 +61,52 @@ extends Diffy[T](ignore, unordered, unorderedFieldKeys) {
       val name = f.getName
       val fullName = if (root.isEmpty) name else root + "." + name
       if (f.isRepeated && unordered.contains(fullName)) {
-        val getFieldFn: Option[AbstractMessage => AnyRef] =
-          unorderedFieldKeys.get(fullName).map(s => getField(getFieldDescriptor(f, s)) _)
-        val a = sortList(x.getField(f).asInstanceOf[java.util.List[AbstractMessage]], getFieldFn)
-        val b = sortList(y.getField(f).asInstanceOf[java.util.List[AbstractMessage]], getFieldFn)
-        if (f.getJavaType == JavaType.MESSAGE && unordered.exists(_.startsWith(s"$fullName."))
-            && unorderedFieldKeys.contains(fullName)) {
-          a.asInstanceOf[java.util.List[AbstractMessage]].asScala.zip(
-            b.asInstanceOf[java.util.List[AbstractMessage]].asScala).flatMap {
-              case (l, r) => diff(l, r, f.getMessageType.getFields.asScala, fullName)}
+        if (f.getJavaType == JavaType.MESSAGE
+          && unordered.contains(fullName)
+          && unorderedFieldKeys.contains(fullName)) {
+          val l = x.flatMap(r =>
+            Option(r.getField(f).asInstanceOf[java.util.List[AbstractMessage]].asScala))
+            .getOrElse(List())
+            .flatMap(m => Try(getFieldDescriptor(f, unorderedFieldKeys(fullName))).toOption
+              .flatMap(fd => getField(fd)(m)).map(k => (k, m))).toMap
+          val r = y.flatMap(r =>
+            Option(r.getField(f).asInstanceOf[java.util.List[AbstractMessage]].asScala))
+            .getOrElse(List())
+            .flatMap(m => Try(getFieldDescriptor(f, unorderedFieldKeys(fullName))).toOption
+              .flatMap(fd => getField(fd)(m)).map(k => (k, m))).toMap
+          (l.keySet ++ r.keySet).flatMap(k =>
+            diff(l.get(k),
+              r.get(k),
+              f.getMessageType.getFields.asScala,
+              fullName))
         }
         else {
-          if (a == b) Nil else Seq(Delta(fullName, Option(a), Option(b), delta(a, b)))
+          val a = x.flatMap(r =>
+            Option(r.getField(f).asInstanceOf[java.util.List[AbstractMessage]])).map(sortList)
+          val b = y.flatMap(r =>
+            Option(r.getField(f).asInstanceOf[java.util.List[AbstractMessage]])).map(sortList)
+          if (a == b) Nil else Seq(Delta(fullName, a, b, delta(a.orNull, b.orNull)))
         }
       } else {
         f.getJavaType match {
           case JavaType.MESSAGE if !f.isRepeated =>
-            val a = getField(f)(x).asInstanceOf[AbstractMessage]
-            val b = getField(f)(y).asInstanceOf[AbstractMessage]
-            if (a == null && b == null) {
+            val a = x.flatMap(m => getField(f)(m).asInstanceOf[Option[AbstractMessage]])
+            val b = y.flatMap(m => getField(f)(m).asInstanceOf[Option[AbstractMessage]])
+            if (a.isEmpty && b.isEmpty) {
```

Comment:
I liked the case statement you changed the other file to use in the same part of that logic a bit better (but this is still ok)